WIP on dependencies (#42)
adoptingerlang/adoptingerlang: adoptingerlang.org

:EXPORT_HUGO_MENU: :menu main :parent development
:END:

> One major distinction between programming 20 years ago and programming today is in how software projects shifted in their approaches. Whereas material of yesterday focused a lot on writing reusable and extendable components within each project, the current trend is to make smaller, isolated projects that each individually ends up being easy to throw away and replace. Current microservice implementations and Javascript dependency trees may lead us to believe no project is too small to be discardable.

Reviewer: "in how software projects shifted in their approaches"

Not sure how to rephrase this but wanted to note it reads funky.

Ironically, this approach towards strong isolation and clearly-defined boundaries—whether they be interfaces, network APIs, or protocols—that was necessary for small single-purpose components has given rise to what is possibly the greatest amount of code reuse we've ever seen through libraries and packages. Reuse materialized from shared needs across various people and projects, not from the mythical ability of providing the perfectly extendable class hierarchy. There is so much reuse today that we have reached the point where folks are starting to ask if maybe, just maybe, we aren't reusing more code than we should: each library comes with risks and liabilities, and we're exposing ourselves to a lot of it for the sake of moving fast.

> Erlang, for its own part, has always lived and breathed strong isolation with clearly defined message-passing protocols (often wrapped behind functional interfaces). Its small community size of generally experienced developers probably made it lag behind other communities when it comes to libraries and packaging. It was not until 2009 that it became easy to install libraries from other people, through the first version of =rebar=. Erlang libraries proliferated all over GitHub and similar pages haphazardly, and at some point, it was possible to find over twelve different versions of the same PostgreSQL driver, which all had the same name and similar versions, but all did a few things differently. Multiple people tried to write package managers that would simplify the installation of dependencies, but none managed take hold.

Reviewer: I think a large part of what made rebar catch on (resulting in it being easy to install libraries from other people) was being from Basho, that and being an escript. So I'd phrase it as "Despite the efforts of CEAN from ProcessOne and faxien, no package manager gained traction in the community, it was not until 2009 that it really became easy to install libraries from other people, through the first version of rebar."

Reviewer: Oh, maybe also a good thing to mention in these first few paragraphs and that rebar brought new to the ecosystem, is the move to per-project dependency installation. That is something that changed across the industry and pretty sure rebar was the first to introduce it to Erlang, while faxien and CEAN installed applications into the OTP lib.

As such, Rebar3 has the following properties:

- Rebar3 is a declarative build tool. You provide a configuration, and it executes code to satisfy that configuration. If you want to run custom scripts and extend the build, it specifically provides [[][hooks]] or [[][plugins]] interfaces to do so.

> - As part of that declarative approach, commands in Rebar3 are defined with dependency sequences. For example, the =rebar3 compile= task relies on =rebar3 get-deps= (or rather, a private form of it), and will run it for you. The command =rebar3 tar= will implicitly call a sequence that includes =get-deps -> compile -> release -> tar= for you.

Author: Will do; get-deps is an alias for `lock` in the first place.

   │  │  └─ wts─0.3.0 (hex package)
   │  │     └─ rfc3339─0.9.0 (hex package)
   │  └─ telemetry─0.4.0 (hex package)
   ├─ grpcbox─0.11.0 (hex package)
   │  ├─ acceptor_pool─1.0.0 (hex package)
   │  ├─ chatterbox─0.9.1 (hex package)
   │  │  └─ hpack─0.2.3 (hex package)
   │  ├─ ctx─0.5.0 (hex package)
   │  └─ gproc─0.8.0 (hex package)
   ├─ jsx─2.10.0 (hex package)
   ├─ pgo─0.8.0+build.91.refaf02392 (git repo)
   │  ├─ backoff─1.1.6 (hex package)
   │  └─ pg_types─0.0.0+build.24.ref32ed140 (git repo)
   ├─ recon─2.4.0 (hex package)
   └─ uuid─1.7.5 (hex package)
      └─ quickrand─1.7.5 (hex package)
#+END_SRC

In this listing, all top-level applications (those we write in the repository) are at the top, and the dependencies we just fetched are all below them. You can see the entire resolution tree that way, find which versions have been fetched, and by which application they have been brought in. This can prove useful when a transitive dependency is imported by two other applications, and only one wins.

If you take a look within =_build/default/lib=, you will see all these applications within their own directory:

#+NAME: service_discovery.deps-ls
#+BEGIN_SRC sh

> #λ service_discovery → master* → ls _build/default/lib

Author: that was some fat fingering.

**** Building a Project with Dependencies

Building actual projects requires more than just fetching its dependencies and compiling them. As mentioned in [[What Makes a Lib an App]], the Erlang run-time system expects to find run-time definitions of dependencies within the =.app= or =.app.src= file. Not putting them there means that while your libraries are available on disk, they won't necessarily be visible during some build phases, some tests, or in releases.

For everything to work, we must add them. Open up =apps/service_discovery/src/service_discovery.app.src= and look at the values in the =applications= tuple:

#+NAME: service_discovery.app-src
#+BEGIN_SRC erlang
{application, service_discovery,
 [{description, "Core functionality for service discovery service"},
  {vsn, {git, short}},
  {registered, []},
  {mod, {service_discovery_app, []}},
  {applications,
   [kernel,
    stdlib,
    erldns,
    service_discovery_storage
   ]},
  {env,[]},
  {modules, []},

  {licenses, ["Apache 2.0"]},
  {links, []}
 ]}.
#+END_SRC

You can see =erldns= and =service_discovery_storage= have been added. Only when such dependencies are specified will everything both build and run fine.

> If you have ever worked with other build tools in the Erlang ecosystem, you likely never had to do this. These tools (erlang.mk or Mix in elixir) end up copying dependencies in the project configuration into the =applications= tuple for you. It sounds like a huge hindrance to have to do this by hand instead, but it ends up giving one critical piece of control in letting you decide _when_ something is used. Let's see why that might be important.

Author: I had no idea this was doable for them.

- you have your main project somewhere on your file system
- the dependency is declared in the rebar.config file of the main project
- you also have that dependency somewhere on your file system, as a standalone project
- you add a =_checkouts/= directory to the main project
- you either copy or symlink the dependency's directory in the =_checkouts/= directory

From that point on, every time the main app is built, it will add an =ebin/= directory to the dependency's directory in checkouts, and re-compile it as if it were a top-level application within the main app.

You can then test the changes to your dependency within the main application until they're ready. Once you're done, remove the =ebin/= directory from the dependency, commit and publish your code in the dependency, remove it from the =_checkouts= directory, and =rebar3 upgrade= it.

This lets you do a lot of small iterative changes locally on a dependency, within the context of the main app, without having to make everything public all the time. It massively reduces the overhead of per-app repositories as an overall development strategy.

*** Using Elixir Dependencies

> For many years, the Elixir and Erlang road was a one way street. You could call Erlang from Elixir, but the opposite wasn't true. Since then, and thanks to the support of the Erlang Ecosystem Foundation, changes have been made to Rebar3 to give it a brand new compiler management structure. This structure has made it possible to write compiler plugins that can let Erlang users use Elixir code.

Reviewer: ```suggestion
For many years, the Elixir and Erlang road was a one way street. You could include Erlang dependencies in your Elixir project, but the opposite wasn't true. Since then, and thanks to the support of the Erlang Ecosystem Foundation, changes have been made to Rebar3 to give it a brand new compiler management structure. This structure has made it possible to write compiler plugins that can let Erlang users use Elixir code.
```

The way to do it is to first install Elixir. For this you might want to follow the steps at LOCATION X. Most Elixir developers use =asdf= as a tool to manage versions. As described in TODO SECTION OF BUILD TOOLS, =kerl= options for Erlang can be used with the Erlang plugin for =asdf=, so that can give you a complete setup.

With Elixir in place, add the following plugin to your library or project:

TODO

From that point on, you will be able to install any hex dependency containing Elixir code without a problem.

> TODO: check if we can use git deps now?

Reviewer: Not yet. Requires merging with that other persons plugin which I haven't done yet.

Reviewer: I guess it should mention that is coming soon and once I get that done we can update it.

> Do note that mixed projects using both Erlang and Elixir within the same library are not currently supported since more work both on the rebar3 and Elixir side would need to be done to make this possible.

Reviewer: ```suggestion
Do note that mixed rebar3 projects using both Erlang and Elixir within the same library are not currently supported since more work both on the rebar3 and Elixir side would need to be done to make this possible.
```

Since this works in mix it is worth making it clear this is about rebar3 and not the languages general, except for the Elixir compiler issues and their not being available as a public api.

*** Corporate Environments

Corporate environments tend to have all kinds of weird restrictions regarding what can or cannot be done, and development tools can be very idiosyncratic. Rebar3 was mostly developed to fit an open-sourced world and focuses mostly on enforcing project structure, fetching dependencies, and using both together as a great pretext to wrap a bunch of standard tools.

> As such, it is perhaps unsurprising that fitting a corporate environment with the tool can prove to be a bit challenging. In this section, we'll cover some standard tools that turn out to be common to multiple corporate environments, and risk making your life easier when adopting Erlang.

Reviewer: ```suggestion
As such, it is perhaps unsurprising that fitting a corporate environment with the tool can prove to be a bit challenging. In this section, we'll cover some standard tools that are common to corporate environments and risk making your life harder when adopting Erlang.
```

Author: Will take only part of this one. I think the tools risk making your life easier, even though the corpo env make it harder? at least that was the intent

Reviewer: Maybe I'm misunderstanding. How would something "risk" making life easier?

Author: Well "there is a chance it will make life easier" I guess would be the logical equivalent. I can switch to that.

Rebar3 supports this use case. If you want to enable it, you will need to first set up a private hex instance.
TODO: fill instructions from what Starbelly had written.
TODO: fill instructions to use minirepo

**** About Monorepos

If you work in a corporate environment with a monorepo, rebar3 is not the best tool for the job. Mostly, this comes down to the fact that most companies using a monorepo have a lot of custom tooling with very custom workflows, large codebases, and a very strong propension to never share access with maintainers of rebar3. Until it becomes possible for maintainers to get access, little will be possible to do on that front.

> Various commercial users who rely on rebar3 despite using monorepos have gotten successful builds with a combination of using =_checkout= dependencies, along with modifying the =_build= directory (TODO: add option)

Reviewer: We don't want to encourage this though, right?

I also think we should make clear an umbrella project is not the same thing as a monorepo.

I have a lot of questions about why rebar3 doesn't work for these monorepos too.. maybe they aren't using releases, I guess. Having a rebar3  project in a monorepo that builds a release for deployment shouldn't be an issue, aside from there not being a Bazel/Buck plugin.
